### DISCUSSION TITLE
Who's thoughts are those?

### DISCUSSION THREAD

**Discussion by iplayfast** [crewAIInc/crewAI#1102] [Feature Requests]
I've got code with two agents, one can delegate to the other.
I'm getting results like
```

> Entering new AgentExecutor chain...

Thought: Do I need to use a tool? Yes
Action: Search the internet - Trends in AI
Action Input: NoneSearch the internet - Trends in AI is not a valid tool, try one of [Search the internet].Thought: Do I need to use a tool? Yes
Action: Search the internet - Trends in AI
Action Input: NoneSearch the internet - Trends in AI is not a valid tool, try one of [Search the internet].
```

It would be very useful to have the Role before the thought, so like
'Writer:Thought"
or
"Researcher:Thought"

**Comment by joaomdmoura** [comment]
good call, specially upon delegation it can get trick to tag along, I'll add it

**Comment by caio-swdev** [comment]
Is my take on solving this issue valid?
It solves it but the disadvantage to this solution is that it added more text to the prompt and so it raises the token count.

**Comment by joaomdmoura** [comment]
Ideally no, as we have that information in the executor we can make sure to log them without having to update the prompts itself

**Comment by jingchang0623-crypto** [comment]
Oh man, this problem goes DEEPER than you think. 😂

I ran a 5-agent crew last month — one researcher, one writer, one coder, one reviewer, one orchestrator. Within the first hour, the "writer" agent started thinking like the researcher, the "reviewer" was just echoing the coder, and the orchestrator had basically become a sentient rubber stamp.

It was like watching a meeting where everyone slowly becomes the same person. Corporate meeting syndrome, but for AI.

**What actually worked for us:**
1. **Prefix every prompt with the role** — `WRITER: ...` / `RESEARCHER: ...` right at the top, not buried somewhere in the middle
2. **Give each agent a "personality fingerprint"** — literally tell the researcher "you're skeptical, always ask but is this source reliable?" and the writer "you care about readability first, accuracy second". Conflicting personalities = clearer outputs
3. **Log who said what** — we pipe every agent's thoughts to a structured JSON file with a role field. When debugging delegation chaos, you can actually trace the blame trail

The whole saga is documented here if you want a laugh: https://miaoquai.com/stories/multi-agent-meeting-hell.html

TL;DR: multi-agent orchestration is basically herding cats. But cats that can also write Python.

**Comment by kinthaiofficial** [comment]
The question of "whose thoughts are those" gets at something fundamental about multi-agent reasoning transparency — when one agent's output informs another, the reasoning chain becomes distributed and hard to attribute.

In a multi-agent system, "thoughts" can come from:

**The current agent's LLM inference** — most visible, usually well-attributed.

**Retrieved memory from previous sessions** — the agent is expressing a belief formed in a past context. If that context was wrong or outdated, the belief is now confidently stale. Attribution: "retrieved from memory, formed on [date]."

**Knowledge inherited from a parent agent's context** — when a root agent primes a sub-agent with a summary of the situation, the sub-agent's reasoning incorporates that framing. If the summary was lossy or biased, the sub-agent inherits the bias. Attribution: "context provided by agent X."

**Pattern learned from training data** — the model has a "prior belief" before any context is added. This one is hardest to attribute and most likely to be wrong in specialized domains.

For agent systems, making thought provenance explicit is operationally useful:
- "I believe X [from my own inference]" — high confidence in the reasoning, may need fact-checking
- "X [retrieved from memory, 3 days old]" — potentially stale, verify if critical
- "X [from parent agent's context summary]" — may have framing bias

Signed provenance at the thought level is what we've been building toward in KinthAI: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons — the audit trail design that makes this traceable.

In your multi-agent setup, do you have a practical need to trace which agent contributed which piece of reasoning, or is this more of a philosophical concern?
